awx/main/scheduler/dag_workflow.py

```diff
@@ -63,40 +112,113 @@ def cancel_node_jobs(self):
+            }
+            if len(failed_path_nodes_id_status) > 0:
+                parms['node_status'] = ",".join(["({},{})".format(id, status) for id, status in failed_path_nodes_id_status])
+            if len(failed_unified_job_template_node_ids) > 0:
+                parms['no_ufjt'] = ",".join(failed_unified_job_template_node_ids)
+            return True, smart_text(s.format(**parms))
+        return False, None
+
+    r'''
+    Determine if all nodes have been decided on being marked do_not_run.
+    Nodes that are do_not_run False may become do_not_run True in the future.
+    We know a do_not_run False node will NOT be marked do_not_run True if there
+    is a job run for that node.
+
+    :param workflow_nodes:     list of workflow_nodes
+
+    Return a boolean
+    '''
+    def _are_all_nodes_dnr_decided(self, workflow_nodes):
+        for n in workflow_nodes:
+            if n.do_not_run is False and not n.job and n.unified_job_template:
+                return False
+        return True
+
+
+    r'''
+    Determine if a node (1) is ready to be marked do_not_run and (2) should
+    be marked do_not_run.
+
+    :param node:             SimpleDAG internal node
+    :param parent_nodes:     list of workflow_nodes
+
+    Return a boolean
+    '''
+    def _should_mark_node_dnr(self, node, parent_nodes):
+        for p in parent_nodes:
+            if p.do_not_run is True:
+                pass
+            elif p.job:
+                if p.job.status == 'successful':
+                    if node in (self.get_dependencies(p, 'success_nodes') +
+                                self.get_dependencies(p, 'always_nodes')):
+                        return False
+                elif p.job.status in ['failed', 'error', 'canceled']:
+                    if node in (self.get_dependencies(p, 'failure_nodes') +
+                                self.get_dependencies(p, 'always_nodes')):
+                        return False
+                else:
+                    return False
+            elif p.do_not_run is False and p.unified_job_template is None:
+                if node in (self.get_dependencies(p, 'failure_nodes') +
+                            self.get_dependencies(p, 'always_nodes')):
+                    return False
             else:
-                # Job is about to run or is running. Hold our horses and wait for
-                # the job to finish. We can't proceed down the graph path until we
-                # have the job result.
-                return False, False
-        return True, is_failed
+                return False
+        return True
+
+    def mark_dnr_nodes(self):
+        root_nodes = self.get_root_nodes()
+        nodes_marked_do_not_run = []
+
+        for node in self.sort_nodes_topological():
+            obj = node['node_object']
+
+            if obj.do_not_run is False and not obj.job and node not in root_nodes:
+                parent_nodes = [p['node_object'] for p in self.get_dependents(obj)]
```

Comment:
If I read "dependents", I think if children, not parents, and that makes it hard for me to follow some parts of this. I know I'm applying laymen terminology to this, but a dependent is someone you claim a deduction for on your taxes. Dependencies sound like parents, but refer to children here. Maybe one day we can rename `get_dependents` to `get_parents`.